spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageResource.java

```diff
@@ -0,0 +1,332 @@
+	 */
+	public URL createSignedUrl(TimeUnit timeUnit, long timePeriods)  {
+		if (LOGGER.isWarnEnabled()) {
+			if (!exists()) {
+				LOGGER.warn("Creating signed URL for non-existing GCS object " + this.location);
+			}
+		}
+
+		return this.storage.signUrl(
+				BlobInfo.newBuilder(getBlobId()).build(), timePeriods, timeUnit);
+	}
+
+	/**
+	 * Creates the blob that this {@link GoogleStorageResource} represents in Google Cloud
+	 * Storage.
+	 * @return the created blob object
+	 * @throws StorageException if any errors during blob creation arise,
+	 * such as if the blob already exists
+	 * @throws IllegalStateException if the resource reference is to a bucket, and not a blob.
+	 */
+	public Blob createBlob() throws StorageException {
+		return this.storage.create(BlobInfo.newBuilder(getBlobId()).build());
+	}
+
+	/**
+	 * Creates the bucket that this resource references in Google Cloud Storage.
+	 * @return the {@link Bucket} object for the bucket
+	 * @throws StorageException if any errors during bucket creation arise,
+	 * such as if the bucket already exists
+	 */
+	public Bucket createBucket() {
+		return this.storage.create(BucketInfo.newBuilder(this.bucketName).build());
+	}
+
+	/**
+	 * Returns the {@link Bucket} associated with the resource.
+	 * @return the bucket if it exists, or null otherwise
+	 */
+	public Bucket getBucket() {
+		return this.storage.get(this.bucketName);
+	}
+
+	/**
+	 * Checks for the existence of the {@link Bucket} associated with the resource.
+	 * @return true if the bucket exists
+	 */
+	public boolean bucketExists() {
+		return getBucket() != null;
+	}
+
+	private Blob throwExceptionForNullBlob(Blob blob) throws IOException {
+		if (blob == null) {
+			throw new FileNotFoundException("The blob was not found: " + this.location);
+		}
+		return blob;
+	}
+
+	@Override
+	public File getFile() throws IOException {
+		throw new UnsupportedOperationException(
+				getDescription() + " cannot be resolved to absolute file path");
+	}
+
+	@Override
+	public long contentLength() throws IOException {
+		return throwExceptionForNullBlob(getBlob()).getSize();
+	}
+
+	@Override
+	public long lastModified() throws IOException {
+		return throwExceptionForNullBlob(getBlob()).getUpdateTime();
+	}
+
+	/**
+	 * Creates a file in the same bucket as this one. It inherits {@code autoCreateFiles}
+	 * from this object.
+	 *
+	 * <p>The file is created if it didn't exist in Google Cloud Storage.
+	 * @param relativePath the URL to a Google Cloud Storage file
+	 * @return the {@link GoogleStorageResource} object for the created file
+	 * @throws IOException
+	 */
+	@Override
+	public GoogleStorageResource createRelative(String relativePath) throws IOException {
+		GoogleStorageResource blobResource = new GoogleStorageResource(
+				this.storage, this.location.resolve(relativePath).toString(), this.autoCreateFiles);
+
+		if (!blobResource.exists()) {
+			blobResource.createBlob();
```

Comment:
Why not just like that relative blob determine its own creation in its own getOutputStream method like the original resource does?

Reply:
As mentioned before in another PR, a user might want to create a blob without writing to it.
When I call `createRelative()`, I'm expecting a blob to be created right away.

Reply:
But you would getOutputStream on that other object, which woudl create it, before you write to it.

And if you were to say you would want to write to it via a signedURL, then I would say the signed URL function has the warning about that exact situation already. I don't see a reason in this asymmetry in behavior between the original resource and relative resources you get from it.

Reply:
Why would you call `getOutputStream()` on the other object? There are no guarantees that will happen.
For all we're concerned, another completely different application which doesn't even use Spring or Java might write to that file.

Reply:
I'm removing this code because judging by the `FileSystemResource` implementation, `createRelative` should not create any resources, but simply provide a handle to the relative resource.